src/controller.js

```diff
@@ -0,0 +1,136 @@
+    };
+
+    renderPorts();
+    renderShip();
+
+    window.setInterval(() => {
+      document.querySelector('#viewport').style.backgroundImage = `url('${backgrounds[backgroundIndex % backgrounds.length]}')`;
+      backgroundIndex += 1;
+    }, 1000);
+  };
+
+  Controller.prototype.renderShip = function renderShip() {
+    const ship = this.ship;
+    const currentPortIndex = ship.itinerary.ports.indexOf(ship.currentPort);
+    const nextPortIndex = currentPortIndex +1;
+
+    const currentPortName = ship.currentPort.name;
+    const nextPortName = (nextPortIndex < ship.itinerary.ports.length)
+      ? ship.itinerary.ports[nextPortIndex].name
+      : 'End of the line';
+
+    const currentPortElement = document.querySelector('#currentPort');
+    currentPortElement.textContent = `Current Port: ${currentPortName}`;
+
+    const nextPortElement = document.querySelector('#nextPort');
+    nextPortElement.textContent = `Next Port: ${nextPortName}`;
+  };
+
+  Controller.prototype.setSail = function setSail() {
+    const ship = this.ship;
+    const currentPortIndex = ship.itinerary.ports.indexOf(ship.currentPort);
+    const nextPortIndex = currentPortIndex +1;
+    const nextPortElement = document.querySelector(`[data-port-index='${nextPortIndex}']`);
+
+    const shipElement = document.querySelector('#ship');
+    const sailInterval = setInterval(() => {
+      const shipLeft = parseInt(shipElement.style.left, 10);
+      if (shipLeft === nextPortElement.offsetLeft - 32) {
+        ship.setSail();
+        ship.dock();
+        clearInterval(this.sailInterval); 
+        this.renderShip();
+
+        this.renderMessage(`Now departing ${ship.currentPort.name}`);
+      }
+      shipElement.style.left = `${shipLeft + 1}px`;
+    }, 20);
+
+    this.sailInterval = sailInterval;
+
+    if (!nextPortElement) {
+      this.renderMessage(`End of the line!`);
+      alert('End of the line!');
+    } else {
+      this.renderMessage(`Now departing ${ship.currentPort.name}`);
+    }
+  };
+
+  Controller.prototype.renderMessage = function renderMessage(message) {
+    const messageElement = document.createElement('div');
+    messageElement.id = 'message';
+    messageElement.innerHTML = message;
+    const viewport = document.querySelector('#viewport');
+    viewport.appendChild(messageElement);
+
+    setTimeout(() => {
+      viewport.removeChild(messageElement);
+    }, 2000);
+  };
+
+  Controller.prototype.initForm = function initForm() {
+    const addPortForm = document.querySelector('#addPortForm');
+    addPortForm.addEventListener('submit', (event) => {
+      event.preventDefault(); 
+      const portNameInput = document.querySelector('#portNameInput');
+      const portName = portNameInput.value;
+      if (portName.trim() !== '') {
+        const newPort = new Port(portName);
+        this.ship.itinerary.addPort(newPort);
+        portNameInput.value = ''; 
+      }
+    });
+  };
```

Comment:
here, you've added functionality to add ports to your `ship.itinerary`, however, you haven't added any way to render these new ports to the screen. if you add a new prototype method:
```
Controller.prototype.renderPorts = function() {
    const portsContainer = document.getElementById('ports');
    while (portsContainer.firstChild) {
      portsContainer.removeChild(portsContainer.firstChild);
    };

    let portsWidth = 0;
    const { ports } = this.ship.itinerary

    ports.forEach((port, i) => {
        const portElement = document.createElement('div');
        portElement.className = 'port';
        portElement.dataset.portName = port.name;
        portElement.dataset.portIndex = i;

        portsWidth += 256;

        portsContainer.appendChild(portElement);
    });

    portsContainer.style.width = `${portsWidth}px`;
}
```
you can then add the line `this.renderPorts()` inside of your if statement, i.e.
```
if (portName.trim() !== '') {
  const newPort = new Port(portName);
  this.ship.itinerary.addPort(newPort);
  portNameInput.value = ''; 
  this.renderPorts();
}
```
your ports will now render on screen when you add them using the form